### ISSUE TITLE
How to use groupby aggregations on categorical columns

### ISSUE THREAD

**Issue by gaganmalhotra** [vaexio/vaex#910]
Hello,

Recently i have been trying to switch over from using pandas to vaex but have stumbled upon a basic issue of using groupby on categorical columns --

For example, we have sample data as - >
```
studentData = {
    'name' : ['jack', 'jack', 'Riti', 'Aadi'],
    'age' : [34, 28, 30, 16],
    'city' : ['Sydney', 'Hong Kong', 'Delhi', 'New york']
}
dfObj = pd.DataFrame(studentData)
```
DOESN'T WORKS with Vaex ---

Converting to vaex and doing a group by - 
```
## calculate card_id level CTR's

vaex_df = vaex.from_pandas(dfObj, copy_index=False)
    
data = vaex_df.groupby(by = ['name'], agg = {
    'age_sum': vaex.agg.sum('age'),
    'age_count': vaex.agg.count('age'),
    'city' : vaex.agg.first('city', 'name'),
})
```

When i try to run the above groupby, there is a following error >
```
ValueError: Could not find a class (AggFirst_string), seems <class 'str'> is not supported
````

WORKS with Pandas ---
Pandas data frame does give the first string `city` if we use max aggregation like below - 
```
pandas_df.groupby(by = ['name']).agg({
    'age': ['sum', 'count'],
    'city' : ['first'],
})
```

Is there a way to make it work, i really like using vaex since its super fast
thank you

**Comment by gaganmalhotra** [comment]
> Hello,
> 
> Recently i have been trying to switch over from using pandas to vaex but have stumbled upon a basic issue of using groupby on categorical columns --
> 
> For example, we have sample data as - >
> 
> ```
> studentData = {
>     'name' : ['jack', 'jack', 'Riti', 'Aadi'],
>     'age' : [34, 28, 30, 16],
>     'city' : ['Sydney', 'Hong Kong', 'Delhi', 'New york']
> }
> dfObj = pd.DataFrame(studentData)
> ```
> 
> **_DOESN'T WORKS with Vaex ---_**
> 
> Converting to vaex and doing a group by -
> 
> ```
> ## calculate card_id level CTR's
> 
> vaex_df = vaex.from_pandas(dfObj, copy_index=False)
>     
> data = vaex_df.groupby(by = ['name'], agg = {
>     'age_sum': vaex.agg.sum('age'),
>     'age_count': vaex.agg.count('age'),
>     'city' : vaex.agg.first('city', 'name'),
> })
> ```
> 
> When i try to run the above groupby, there is a following error >
> 
> ```
> ValueError: Could not find a class (AggFirst_string), seems <class 'str'> is not supported
> ```
> 
> _**WORKS with Pandas ---**_
> Pandas data frame does give the first string `city` if we use max aggregation like below -
> 
> ```
> pandas_df.groupby(by = ['name']).agg({
>     'age': ['sum', 'count'],
>     'city' : ['first'],
> })
> ```
> 
> Is there a way to make it work, i really like using vaex since its super fast
> thank you

@maartenbreddels any clues on this?

**Comment by JovanVeljanoski** [comment]
Hi  @gaganmalhotra 

`vaex.agg.first` does not work on string columns. Here I would encode the string columns as ordinal, and than it is easy to do what you need. So your example would look something like this:

```
studentData = {
    'name' : ['jack', 'jack', 'Riti', 'Aadi'],
    'age' : [34, 28, 30, 16],
    'city' : ['Sydney', 'Hong Kong', 'Delhi', 'New york']
}
dfObj = pd.DataFrame(studentData)

df = vaex.from_pandas(dfObj)

df = df.ordinal_encode(column='name')
df = df.ordinal_encode(column='city')

group = df.groupby(by = ['name'], agg = {
    'age_sum': vaex.agg.sum('age'),
    'age_count': vaex.agg.count('age'),
    'city' : vaex.agg.first('city', 'name'),
})

```

To be able to map back from indices to the actual elements, you can see the dictionary `df._categories`. 

Alternatively you can use `vaex.ml.LabelEncoder()` to encode the strings to ints, if you care about ML workflows. The example above is more suitable for quick experimentation with the data.

I hope this helps.

**Comment by gaganmalhotra** [comment]
Interesting! 
thanks @JovanVeljanoski 
ill follow labelEncoder() for my workflow

**Comment by maartenbreddels** [comment]
I just renamed the issue, this seems to be a missing feature, thanks for the report.

**Comment by dclong** [comment]
It seems that other aggregation functions (e.g., max, min) are not supported on string columns either. Those are definitely missing features.

**Comment by JovanVeljanoski** [comment]
What is the `max` of a bunch of strings?

**Comment by dclong** [comment]
Just compare using string's nature ordering. It is not rare. Sometimes people use strings for dates and timestamps. To get the max/min date/timestamp, they essentially do a string comparisons.

**Comment by maartenbreddels** [comment]
Also, as it is now, you can get the first X based on y, like the first
country, sorted by GDP, grouped by continent. I don't think we expose all
those combinations.

(from mobile phone)

Op ma 4 okt. 2021 21:48 schreef Ben Chuanlong Du ***@***.***>:

**Comment by tfmark** [comment]
Is there a technical reason behind `first`/`last` aggregation types not being supported for strings - is it tricky to implement? I'm happy to look into raising a PR if there's not a fundamental thing preventing it from working :)

**Comment by maartenbreddels** [comment]
No, it's an oversight, we simply did not implement that. It requires C++ skills, would you willing to go there? If so, that's also a good reason to do some refactoring to make it easier to implement it maybe.

**Comment by tfmark** [comment]
It's definitely outside my comfort zone... I've been looking at the AggFirst code in [superagg.cpp](https://github.com/vaexio/vaex/blob/732c4e93e3237db9b14c338a225e4f6cf573c196/packages/vaex-core/src/superagg.cpp#L437) and it's not *clicked* just yet. I don't get what `grid_data_order` is for - why aren't you taking the first item you find?

I'll try and clone the repo locally and see if I can get the tests to pass, and then throw some printf statements around to see if I can get an idea of how this is working.

**Comment by maartenbreddels** [comment]
it's maybe a bit different then what you expect, it's the first sorted by a second column https://vaex.io/docs/api.html#vaex.dataframe.DataFrame.first

**Comment by maartenbreddels** [comment]
Eventually we want to have both, also first ordered by say 'row number'
